### ISSUE TITLE
[BUG] DKG signatures are not currently compatible with prefixed ecrecover

### ISSUE THREAD

**Issue by drewstone** [webb-tools/protocol-solidity#108]
# Overview
We are using prefixed signatures in Solidity to verify ECDSA signatures from the DKG. The DKG is **_most likely_** not using this prefix as it's Ethereum specific and the DKG is not directly Ethereum related. Therefore, we need to prefix the data we send to the DKG for signature with the hash of the prefixed and hashed data.

## Tasks
- [ ] Update the proposal data being sent for signature to the DKG to include the prefixed string
- [ ] Verify the DKG signatures using tests w/ DKG simulations.

## Tools
- Raw signatures without prefix: https://www.npmjs.com/package/@polkadot/x-noble-secp256k1
- ToSignedMessageHash (w/ prefix): https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/ECDSA.sol#L200

**Comment by drewstone** [comment]
## Proposed solution
If we are dealing with an EVM chain in `dkg-substrate`, which should be handled somehow, then what we need to do is the following:
- Before the proposal is inserted into the `UnsignedProposalQueue` we need to `ECDSA.toEthSignedMessageHash(keccak256(proposal))`. @saiakilesh please double-check this.

This is akin to (rust-like pseudocode)
```rust
if T::ChainId::from(chain_id) is an EVM chain {
    let unsigned_evm_proposal: T::Proposal = vec![...];
    let proposal_hash = keccak256(unsigned_evm_proposal);
    let prefix = b"\x19Ethereum Signed Message:\n32";
    let prefixed_unsigned_evm_proposal = abi.encodePacked(prefix, proposal_hash);
    UnsignedProposalQueue::<T>::insert(
        T::ChainId::from(chain_id),
        DKGPayloadKey::XYZProposal(nonce),
        prefixed_unsigned_evm_proposal.clone(),
    );
}
```

**Comment by saiakilesh** [comment]
Perhaps some clarifying comments: @drewstone could you please read this over and let me know if it looks correct?

Why are things currently not working?

Let’s take the update wrapping fee proposal as a concrete example.

**On the DKG side:**

The data for this proposal looks like:  `originalData = (resourceId, functionSig, nonce, newFee)`, which is 32 + 4 + 4 + 1 = 41 bytes. 

The DKG takes this proposal, hashes it, and then signs the resulting hash. That is, it signs `keccak256(originalData)`.

**On the Solidity side:**

Now, in the Solidity contracts, the following happens:

Let `prefix = "\x19Ethereum Signed Message:\n32”`. We check that `keccak256(abi.encodePacked(prefix, keccak256(originalData)))` is what is signed by the DKG. Clearly, this is different from `keccak256(originalData)` so there is a mismatch here.

**How we can fix it:**

The `newData` for the DKG to sign should instead be `abi.encodePacked(prefix, keccak256(originalData))`. Then, it will sign `keccak256(abi.encodePacked(prefix, keccak256(originalData)))`.

**Comment by saiakilesh** [comment]
> ## Proposed solution
> * Before the proposal is inserted into the `UnsignedProposalQueue` we need to `ECDSA.toEthSignedMessageHash(keccak256(proposal))`. @saiakilesh please double-check this.
>   
>   * https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/ECDSA.sol#L200
> 

ECDSA.toEthSignedMessageHash(keccak256(proposal)) = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(proposal)))

But if my understanding is correct, what we want need to add to the UnsignedProposalQueue is just abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(proposal)).

**Comment by drewstone** [comment]
I'm also thinking in order to parse out what kind of `ChainIdType` we're dealing with that we should augment the `resourceId` parameter.

**Comment by dutterbutter** [comment]
Closed in [#149](https://github.com/webb-tools/dkg-substrate/pull/149)
